### PR TITLE
Python backend

### PR DESCRIPTION
**Pull request by anooj-gandham**

### CODE REVIEW

File: core/models.py [hunk 1]
```diff
@@ -28,7 +30,21 @@ def create_for_url(self, url, referer=None):
...
+            buf.write(open('output','rb').read())
+        else:
+            buf.write(response.content)
+            tempImage = open('output','wb')
+            tempImage.write(response.content)
+            tempImage.close()
+
+        dims = PILImage.open('output').size
+        if dims[0] < 45 or dims[1] < 45:
+            ##Write here logic for clean exit if image is small like icons.
```
Comment (mrraghur) [inline core/models.py]: How did you choose 45x45 to be size of the icons?

File: core/models.py [hunk 2]
```diff
@@ -91,6 +107,7 @@ class Pin(models.Model):
     image = models.ForeignKey(Image, related_name='pin', on_delete=models.CASCADE)
     published = models.DateTimeField(auto_now_add=True)
     tags = TaggableManager()
+    stars = models.CharField(default=0, max_length=32, blank=True, null=True)
```
Comment (mrraghur) [inline core/models.py:114]: How did you arrive at 32?

File: pinry/settings/base.py [hunk 3]
```diff
@@ -22,7 +22,13 @@
     'core',
     'users',
     'pinry_plugins.apps.PinryPluginsConfig',
+    #New Changes
```
Comment (mrraghur) [inline pinry/settings/base.py:25]: Unnecessary comment

File: pinry/settings/base.py [hunk 4]
```diff
@@ -153,10 +159,15 @@
...
+    ##New Changes
+    'DEFAULT_AUTHENTICATION_CLASSES': (
+       'rest_framework.authentication.TokenAuthentication',
+   ),
+    ##New Changes
     'DEFAULT_PERMISSION_CLASSES': [
-        'rest_framework.permissions.IsAuthenticatedOrReadOnly'
+        'rest_framework.permissions.IsAdminUser'
     ],
+    'DEFAULT_SCHEMA_CLASS': 'rest_framework.schemas.coreapi.AutoSchema',
```
Comment (mrraghur) [inline pinry/settings/base.py:172]: Why do we need this?

File: requirements.txt [hunk 5]
```diff
@@ -0,0 +1,112 @@
+antiorm==1.2.1
+appdirs==1.4.4
+attrs==21.2.0
+Automat==20.2.0
+backports.entry-points-selectable==1.1.0
+black==19.10b0
+cairocffi==1.2.0
+CairoSVG==2.5.2
```
Comment (mrraghur) [inline requirements.txt:11]: Do we really need everything in this list?

File: scrapeGithub/.vscode/settings.json [hunk 6]
```diff
@@ -0,0 +1,6 @@
+{
+    "python.pythonPath": "/home/anooj/codes/dev1/gsenv/bin/python3.8",
```
Comment (mrraghur) [inline scrapeGithub/.vscode/settings.json]: Remove this file. It is only for vscode users.

File: scrapeGithub/database.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    for u in urls:
+        urlCheck = connection.execute(f'SELECT url FROM urls_to_be_scraped WHERE url=="{u}"').fetchall()
+        if urlCheck == []:
+            urlsToBeAdded.append(u)
+
+    records = ",".join(list(map(parseRecord,urlsToBeAdded, [False]*len(urlsToBeAdded))))
+    insertCommand = f"INSERT INTO urls_to_be_scraped (url, isScraped) VALUES {records}"
+    connection.execute(insertCommand)
+
+def updateUrlAsScraped(connection,url):
```
Comment (mrraghur) [inline scrapeGithub/database.py]: "markUrlAsScraped"

File: scrapeGithub/database.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+# urls_to_be_scraped = list(map(lambda x:x[0],getAllUrlsToBeScraped(con)))
+# print(urls_to_be_scraped)
+
+# urls = ['https://github.com/anooj-gandham/blogSearch','https://github.com/anooj-gandham/wappalyzer_1','https://github.com/0x09AL','https://github.com']
+# addUrlsToBeScraped(con,urls)
+
+# records = getRecords(con)
+# print(records.fetchall())
+pdb.set_trace()
```
Comment (mrraghur) [inline scrapeGithub/database.py]: Delete commented lines. remove pdb.
What is the purpose of this file?

File: scrapeGithub/githubUrls.txt [hunk 9]
```diff
@@ -0,0 +1,17282 @@
...
+https://github.com/delannoyk/GIFRefreshControl
+https://github.com/delannoyk/KDEDateLabel
+https://github.com/delano/scruffy
+https://github.com/delay
+https://github.com/delay/feathers-vue-blog-admin-demo
+https://github.com/delba/JASON
+https://github.com/delba/Log
+https://github.com/delba/Permission
+https://github.com/delba/SwiftyOAuth
+https://github.com/d
```
Comment (mrraghur) [inline scrapeGithub/githubUrls.txt:17282]: This file data not code. It need not be in repository.

File: scrapeGithub/scrapeGithub/items.py [hunk 10]
```diff
@@ -0,0 +1,12 @@
+# Define here the models for your scraped items
+#
+# See documentation in:
+# https://docs.scrapy.org/en/latest/topics/items.html
+
+import scrapy
+
+
+class ScrapegithubItem(scrapy.Item):
```
Comment (mrraghur) [inline scrapeGithub/scrapeGithub/items.py]: Why do we need this file? Seems like no real code here.

File: scrapeGithub/scrapeGithub/settings.py [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+
+
+# Crawl responsibly by identifying yourself (and your website) on the user-agent
+#USER_AGENT = 'scrapeGithub (+http://www.yourdomain.com)'
+
+# Obey robots.txt rules
+ROBOTSTXT_OBEY = False
+
+# Configure maximum concurrent requests performed by Scrapy (default: 16)
+#CONCURRENT_REQUESTS = 32
```
Comment (mrraghur) [inline scrapeGithub/scrapeGithub/settings.py]: Keep this as 3 or 4

File: scrapeGithub/scrapeGithub/spiders/scrapeImages.py [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+            if readme == []:
+                return
+            readme = readme.getall()[0]
+            imageTags = []
+            for tag in response.xpath('//a[@class="topic-tag topic-tag-link"]/text()').getall(): #Get tags from about me in github repo
+                imageTags.append(tag.split()[0])
+            stars = response.xpath('//a[@class="social-count js-social-count"]/text()').getall()[0].split()[0]
+            imageUrls = etree.HTML(readme).xpath('//img/@src') #Check for <img> tags and get src attribute from them
+            headers = {
+                'Authorization': 'Token fa965469e96d536925bfffa38a64d0919c511713',
```
Comment (mrraghur) [inline scrapeGithub/scrapeGithub/spiders/scrapeImages.py]: How did you get this token? It should not be in public facing code!

File: scrapeGithub/scrapeGithub/spiders/scrapeImages.py [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+                'Content-Type': 'application/json',
+                    }
+            postApi = 'http://localhost:8000/api/v2/pins/'
+
+            if imageUrls != []:
+                imageUrls = list(filter(None,imageUrls)) #Remove empty urls
+                imageUrls = list(map(self.completeUrl,imageUrls)) #Complete the urls if relative urls are given
+                for img in imageUrls:
+                    data = {'url':img,'referer':None,'description':'','tags':imageTags, 'stars':stars}
```
Comment (mrraghur) [inline scrapeGithub/scrapeGithub/spiders/scrapeImages.py]: I don't like description being empty. We should do something about it

File: scrapeGithub/scrapeGithub/spiders/scrapeImages.py [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+            conn = connectToDatabase("../db.sqlite3")
+             
+            addUrlsToBeScraped(conn,githubRepos)
+            updateUrlAsScraped(conn,response.url)
+            pdb.set_trace()
+
+            conn.commit()
+            conn.close()
+        except:
+            pdb.set_trace()
```
Comment (mrraghur) [inline scrapeGithub/scrapeGithub/spiders/scrapeImages.py:112]: Add logging. Remove pdb here and on line 72

File: core/models.py (hunk 1 above)
Comment (anooj-gandham) [reply core/models.py]: I saw many images and came to conclusion, but the "many" may not be sufficient will need to build a better metric for this classification.

File: core/models.py (hunk 2 above)
Comment (anooj-gandham) [reply core/models.py:114]: stars are generally a number represented as string in github Eg: 700, 1k, 2.5k, 7.7k etc.

File: pinry/settings/base.py (hunk 3 above)
Comment (anooj-gandham) [reply pinry/settings/base.py:25]: Ok.

File: pinry/settings/base.py (hunk 4 above)
Comment (anooj-gandham) [reply pinry/settings/base.py:172]: It was default, I have not added this.

File: requirements.txt (hunk 5 above)
Comment (anooj-gandham) [reply requirements.txt:11]: we need cairoSVG, I think cairocffi may be a dependency to this library.

File: scrapeGithub/.vscode/settings.json (hunk 6 above)
Comment (anooj-gandham) [reply scrapeGithub/.vscode/settings.json]: .vscode will be in .gitignore, it was present here as I copied the scraper for another repo, deleted it.

File: scrapeGithub/scrapeGithub/items.py (hunk 10 above)
Comment (anooj-gandham) [reply scrapeGithub/scrapeGithub/items.py]: Deleted.

File: scrapeGithub/scrapeGithub/settings.py (hunk 11 above)
Comment (anooj-gandham) [reply scrapeGithub/scrapeGithub/settings.py]: Ok.

File: scrapeGithub/scrapeGithub/spiders/scrapeImages.py (hunk 12 above)
Comment (anooj-gandham) [reply scrapeGithub/scrapeGithub/spiders/scrapeImages.py]: The process for getting the token is present in the link in the comment above.
This is a development code, in production we will use an environment variable for token.

File: scrapeGithub/scrapeGithub/spiders/scrapeImages.py (hunk 13 above)
Comment (anooj-gandham) [reply scrapeGithub/scrapeGithub/spiders/scrapeImages.py]: I have added description later today after the commit. Its present now.

File: scrapeGithub/scrapeGithub/spiders/scrapeImages.py (hunk 14 above)
Comment (anooj-gandham) [reply scrapeGithub/scrapeGithub/spiders/scrapeImages.py:112]: Sure.

File: scrapeGithub/database.py (hunk 7 above)
Comment (anooj-gandham) [reply scrapeGithub/database.py]: Ok.

File: scrapeGithub/database.py (hunk 8 above)
Comment (anooj-gandham) [reply scrapeGithub/database.py]: Yes, will clean the code.

File: scrapeGithub/githubUrls.txt (hunk 9 above)
Comment (anooj-gandham) [reply scrapeGithub/githubUrls.txt:17282]: Ok.

File: scrapeGithub/inserUrls.py [hunk 15]
```diff
@@ -0,0 +1,22 @@
+from database import connectToDatabase, getAllUrlsToBeScraped, addUrlsToBeScraped, markUrlAsScraped
+import sys
+import random
+
```
Comment (mrraghur) [inline scrapeGithub/inserUrls.py]: File name has missing 't'

File: README.md [hunk 16]
```diff
@@ -17,25 +16,48 @@ For more information ( screenshots and document ) visit [getpinry.com](https://g
...
 
-## Development
+To view the urls in database
+`SELECT url,isScraped FROM urls_to_be_scraped`
 
-See our full documentation at [https://docs.getpinry.com/development/](https://docs.getpinry.com/development/)
+To run Nginx
 
-## Contributors
+- copy pinry.conf from the project directory to /etc/nginx/conf.d/
```
Comment (mrraghur) [inline README.md:50]: I think the file in repo is nginx.conf

File: core/models.py (hunk 2 above)
Comment (mrraghur) [reply core/models.py:114]: Got it

File: scrapeGithub/inserUrls.py (hunk 15 above)
Comment (anooj-gandham) [reply scrapeGithub/inserUrls.py]: I noticed that long ago, still left it like that 😅

File: pinry/settings/base.py (hunk 4 above)
Comment (mrraghur) [reply pinry/settings/base.py:172]: ok

File: requirements.txt (hunk 5 above)
Comment (mrraghur) [reply requirements.txt:11]: ok

File: scrapeGithub/database.py [hunk 17]
```diff
@@ -0,0 +1,56 @@
+import sqlite3
+import pdb
+
```
Comment (mrraghur) [inline scrapeGithub/database.py:3]: Please add comments what this file is for and how to use it. I am assuming these are some utility methods

File: scrapeGithub/database.py [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+def connectToDatabase(database):
+    conn = sqlite3.connect(database, timeout=10)
+    createTable = """
+    CREATE TABLE IF NOT EXISTS "urls_to_be_scraped" (
+	"id"	INTEGER PRIMARY KEY AUTOINCREMENT,
+	"url"	TEXT NOT NULL UNIQUE,
+	"isScraped"	INTEGER DEFAULT 0,
+    "isValidReadme" INTEGER DEFAULT 1,
```
Comment (mrraghur) [inline scrapeGithub/database.py:12]: Add a comment explaining when a readme can be invalid

File: scrapeGithub/scrapeGithub/spiders/scrapeImages.py (hunk 14 above)
Comment (mrraghur) [reply scrapeGithub/scrapeGithub/spiders/scrapeImages.py:112]: Remove pdb from line 112. Add logging

**Review by mrraghur — COMMENTED** [review]
Remove nohup.out file. May be one of data_files.zip or latest.zip also?

Remove file scrapeGithub/scrapy.log 
What is in static.zip file? Do we need it in repo?

File: scrapeGithub/scrapeGithub/spiders/scrapeImages.py (hunk 14 above)
Comment (anooj-gandham) [reply scrapeGithub/scrapeGithub/spiders/scrapeImages.py:112]: I don't know how to add logging.

File: README.md (hunk 16 above)
Comment (anooj-gandham) [reply README.md:50]: Ok.

File: scrapeGithub/database.py (hunk 17 above)
Comment (anooj-gandham) [reply scrapeGithub/database.py:3]: Ok.

File: scrapeGithub/database.py (hunk 18 above)
Comment (anooj-gandham) [reply scrapeGithub/database.py:12]: Not sure, already have added that in TODO.

**Pull request merged**